flink/src/main/java/org/apache/iceberg/flink/source/StreamingReaderOperator.java

```diff
@@ -0,0 +1,220 @@
+ * processing the checkpoint barriers, thus removing any potential back-pressure.
+ */
+public class StreamingReaderOperator extends AbstractStreamOperator<RowData>
+    implements OneInputStreamOperator<FlinkInputSplit, RowData> {
+
+  private static final long serialVersionUID = 1L;
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingReaderOperator.class);
+  private final MailboxExecutorImpl executor;
+  private FlinkInputFormat format;
+
+  private transient SourceFunction.SourceContext<RowData> readerContext;
+
+  private transient ListState<FlinkInputSplit> checkpointState;
+  private transient Queue<FlinkInputSplit> splits;
+
+  private StreamingReaderOperator(
+      FlinkInputFormat format, ProcessingTimeService timeService, MailboxExecutor mailboxExecutor) {
+    this.format = Preconditions.checkNotNull(format, "The InputFormat should not be null.");
+    this.processingTimeService = timeService;
+    this.executor = (MailboxExecutorImpl) Preconditions.checkNotNull(
+        mailboxExecutor, "The mailboxExecutor should not be null.");
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    super.initializeState(context);
+
+    checkpointState = context.getOperatorStateStore().getListState(
+        new ListStateDescriptor<>("splits", new JavaSerializer<>()));
+
+    int subtaskIdx = getRuntimeContext().getIndexOfThisSubtask();
+    splits = Lists.newLinkedList();
+    if (context.isRestored()) {
+      LOG.info("Restoring state for the {} (taskIdx={}).", getClass().getSimpleName(), subtaskIdx);
+
+      for (FlinkInputSplit split : checkpointState.get()) {
+        splits.add(split);
+      }
+    }
+
+    this.readerContext = StreamSourceContexts.getSourceContext(
+        getOperatorConfig().getTimeCharacteristic(),
+        getProcessingTimeService(),
+        new Object(), // no actual locking needed
+        getContainingTask().getStreamStatusMaintainer(),
+        output,
+        getRuntimeContext().getExecutionConfig().getAutoWatermarkInterval(),
+        -1);
+
+    if (!splits.isEmpty()) {
+      enqueueProcessSplits();
+    }
+  }
+
+  @Override
+  public void processElement(StreamRecord<FlinkInputSplit> element) {
+    splits.offer(element.getValue());
+    enqueueProcessSplits();
+  }
+
+  private void enqueueProcessSplits() {
+    executor.execute(this::processSplits, this.getClass().getSimpleName());
+  }
+
+  private void processSplits() throws IOException {
+    do {
+      FlinkInputSplit split = splits.poll();
+      if (split == null) {
+        return;
+      }
+
+      LOG.debug("load split: {}", split);
+      format.open(split);
+
+      RowData nextElement = null;
+      while (!format.reachedEnd()) {
+        nextElement = format.nextRecord(nextElement);
+        if (nextElement != null) {
+          readerContext.collect(nextElement);
```

Comment:
Just to be sure, `collect` is the right method to call to emit a record? That seems oddly named to me.

Reply:
Yes, it's the right method to emit record.  Pls see the javadoc here: https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/SourceFunction.java#L200